### ISSUE TITLE
make default iterations, chains etc configurable

### ISSUE THREAD

**Issue by schmettow** [stan-dev/rstanarm#118]
#### Summary:
rstanarm always does generous 4 x 2000 MCMC iterations by default. These defaults cannot be altered.

#### Description:
A method to change the default does not seem to exist. Users quickly find themselves typing chain = 1, iter = 500 multiple times, when evaluating models.

#### Reproducible Steps:
None of the two established methods below works:

```
library(dplyr)
library(rstanarm)
d <- data_frame(x = runif(10, 0, 1), y = rnorm(10, x, .2))

formals(stan_glm)$iter = 300
formals(stan_glm)$chains = 1
m <- stan_glm(y ~ x, data = d)

rstan::rstan_options(iter = 300)
rstan::rstan_options(chains = 1)
m <- stan_glm(y ~ x, data = d)
```
#### RStanARM Version:
2.12.1

#### R Version:
3.3.1

#### Operating System:
Windows 7/10

**Comment by jgabry** [comment]
Thanks for the suggestion. This is the same in the rstan package itself. A
while back I suggested allowing the default iter and chains for an R
session to be changed using the rstan_options function in rstan, and I may
have even submitted a pull request that we haven't merged. Possible because
there might have been some resistance from other Stan developers on this
point (I forget to be honest), because, although there's nothing special
about exactly 2000 iter and 4 chains, it's a default that we feel
comfortable won't lead people astray in most cases.

But this is certainly a fair topic to raise for discussion again. We've
also had discussion about allowing Stan models to be run for an amount for
time rather than a number of iterations, but that will require changes to
Stan itself that aren't ready quite yet.

Also, when you say that users find themselves typing it chains=1,iter=500
multiple times, do you actually mean chains=1? I think 4 chains is the
minimum number of chains you want to run. For rstanarm models you probably
don't need more than that very often, but I would definitely caution
against running fewer than 4 chains.

Jonah

**Comment by schmettow** [comment]
Regarding the number of chains, I recall a statement in either Kruschke or MacElreath, that you should run multiple chains when evaluating a model, but a single final. However, my specific scenario is debugging a book chapter (markdown), where multiple models are being fit. It would be convenient to speed up the process by short-circuiting the model fitting parts.

As to the solution, I would prefer the standard R way with formals(). To my mind, that just requires to make iter, chain and the like explicit arguments. Now they are in the dots. I recall a related open issue that explicit arguments is better style.

**Comment by jgabry** [comment]
Thanks for following up. A few comments/questions:

On Thursday, September 22, 2016, schmettow <notifications@github.com> wrote:

> Regarding the number of chains, I recall a statement in either Kruschke or
> MacElreath, that you should run multiple chains when evaluating a model,
> but a single final.
>
I hope they didn't say that! If they do, could you point me to where if
possible? That's just a way to give yourself fewer opportunities to detect
problems. Especially now with easy parallelization, I can't think of a
reason not to always run multiple chains. One chain will be a good idea
when we figure out how to run it for infinitely many iterations in a small
finite amount of time ;)

> However, my specific scenario is debugging a book chapter (markdown),
> where multiple models are being fit. It would be convenient to speed up the
> process by short-circuiting the model fitting parts.
>
Yeah that makes sense for debugging purposes.

> As to the solution, I would prefer the standard R way with formals(). To
> my mind, that just requires to make iter, chain and the like explicit
> arguments. Now they are in the dots. I recall a related open issue that
> explicit arguments is better style.
>
Ok yeah, I would be open to adding those as explicit arguments to the
rstanarm modeling functions. That original issue about the dots was about
rstanarm functions that call other rstanarm functions (e.g. stan_glmer.nb
is a wrapper for stan_glmer), whereas in this case the dots are passed to
functions in a different package (rstan), but it's similar and I do see
your point about being able to use formals().

Jonah

>

**Comment by schmettow** [comment]
MacElreath, Statistical Rethinking (2016), p.256:

> 8.4.2 How many chains do you need?
[...] There are three answers to this question: First, when debugging a model use a single chain. Then, when deciding whether the chains are valid, you need more than one chain. Third, when you begin the final run that you'll make inferences from, you only really need one chain. 

To my own (unsystematic) observation, this seems to make sense, as for some tricky models (e.g., a non-linear mixed effects model with Gamma residuals), convergence seems to strongly improve with longer warm-ups. One chain with 3000 warmups and 3000 samples converges much better than three with 1000/1000. My impression is that warmup with Stan matters much more than burn-in with, say, JAGS. MacElreath clarifies this on the same page, and I was at the time very thankful for the insight.

**Comment by jgabry** [comment]
Thanks a lot for finding that passage and sharing. I think that advice is
unfortunate, although I do agree to some extent. More details below.

On Monday, September 26, 2016, schmettow <notifications@github.com> wrote:

> MacElreath, Statistical Rethinking (2016), p.256:
>
> 8.4.2 How many chains do you need?
> [...] There are three answers to this question: First, when debugging a
> model use a single chain. Then, when deciding whether the chains are valid,
> you need more than one chain. Third, when you begin the final run that
> you'll make inferences from, you only really need one chain.
>
> For the debugging part I agree, and even for the rest I think this advice
will typically be ok. But I still strongly advise against it for most
users, and especially for the intended audience of the Rethinking book,
which as far as I know is primarily people learning how to check MCMC
convergence for this first time.

> To my own (unsystematic) observation, this seems to make sense, as for
> some tricky models (e.g., a non-linear mixed effects model with Gamma
> residuals), convergence seems to strongly improve with longer warm-ups. One
> chain with 3000 warmups and 3000 samples converges much better than three
> with 1000/1000.
>
It's true that some models can require more iterations per chain, but if
you're still in the stage of evaluating convergence it's never a good idea
to run one chain. Even three is less than we recommend.

If you have a simple model (like most of the rstanarm models), run at least
4 chains, and see that everything looks great, then I think Richard's
advice can be ok, and it could be fine to run one chain for your final
inferences. But for nontrivial models (and it also depends on the data) any
additional chain you can run gives you a chance to detect pathological
behavior that wasn't identified by the other chains. I'm pretty sure none
of the other Stan developers would disagrees with me on that. Plus, easy
parallelization of the chains makes this much easier than it used to be!

**Comment by schmettow** [comment]
Thanks for the advice. Maybe, in the Rstan manual you could be more explicit about recommended number of chains. You seem to suggest that four is a minimum. I suspect that many people still stick with three on a quad core, to have one core spared for checking their email.

**Comment by jgabry** [comment]
I didn't mean to imply that there's anything fundamentally special about 4,
just that we've found it to be a useful default for the most part. There's
certainly some degree arbitrariness in our defaults, and I think there are
arguments for changing them that are at least worth entertaining, but
although they are imperfect they are based on the collective experience of
the Stan developers and we try to usually err on the side of caution. That
said, I definitely do understand the desire to keep one core free if you
want to do other things on the same machine!

**Comment by avehtari** [comment]
Running anything CPU intensive with all the available cores should not slow down email reading as the operating system you are using is capable of multitasking and email checking does not require much CPU cycles. If you experience slow email checking when running four chains, it's more likely due to running out of memory. If you want do other CPU intensive thing on the same machine you should give lower priority for Stan processes instead of keeping one core free just in case (and in fact if you are running 3 chains in 4 your machine, it's very likely that all 4 cores are used 75% of time for Stan, ie. Stan processes are not locked to a single cores).

**Comment by avehtari** [comment]
> One chain with 3000 warmups and 3000 samples converges much better than three with 1000/1000. 

How did you evaluate the convergence? You can get lower Rhat with one chain if it is not exploring the posterior well, and higher Rhat with several chains as they are started with different initial values and then slow exploration is more easily detected. In order to check this you should run several 3000/3000 chains, too :)

**Comment by jgabry** [comment]
Thanks for chiming in Aki. That's a good point about CPU vs memory.

**Comment by bgoodri** [comment]
We don't want the vast majority of rstan{arm} users to deviate from the default values, much less change the default values in the code. I view default values as things that are safe, rather than things that are optimal.

Although you don't "need" multiple chains to do the final inference and there is some room for bias if you do inference with many short chains, using a few chains for inference is not going to hurt, and it will help if it allows you to detect a problem that you wouldn't have detected with only one chain. In the era of multicore computers, there is not much harm in running multiple chains. For a lot of rstanarm models, 1000 warmup iterations is more than is necessary but it is safe and occasionally necessary (depending on the data).

**Comment by betanalpha** [comment]
>  I view default values as things that are safe, rather than things that are optimal.

+1

**Comment by schmettow** [comment]
I have no systematic observations on the issue. Most of the time 4x2000 works fine, indeed. Replace "check email" with "work on several projects in parallel to bridge MCMC time" ;)
